src/bootstrap/config.rs

```diff
@@ -765,13 +765,14 @@ impl Config {
 
         // set by build.rs
         config.build = TargetSelection::from_user(&env!("BUILD_TRIPLE"));
-        let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+        let base_path =
+            output(Command::new("git").arg("rev-parse").arg("--show-toplevel")).trim().to_owned();
         // Undo `src/bootstrap`
-        config.src = manifest_dir.parent().unwrap().parent().unwrap().to_owned();
+        config.src = PathBuf::from(base_path);
         config.out = PathBuf::from("build");
 
-        config.initial_cargo = PathBuf::from(env!("CARGO"));
-        config.initial_rustc = PathBuf::from(env!("RUSTC"));
+        config.initial_rustc = config.out.join(config.build.triple).join("stage0/bin/rustc");
+        config.initial_cargo = config.out.join(config.build.triple).join("stage0/bin/cargo");
```

Comment:
This is still in `default-opts`, so `config.out` is always `build`. It needs to be set later, after `config.out` is set to the runtime value: https://github.com/rust-lang/rust/blob/db896fd318b1fea977a65d01d030b827d2a5b03f/src/bootstrap/config.rs#L862-L868